File: java/P0019.java
class P0019{

    public static void main(String[] args) {
        ListNode head = new ListNode(1);
        System.out.println(new Solution().removeNthFromEnd(head,1));
    }

  public class ListNode {
      int val;
      ListNode next;
      ListNode() {}
      ListNode(int val) { this.val = val; }
      ListNode(int val, ListNode next) { this.val = val; this.next = next; }
  }

    static class Solution {
        public ListNode removeNthFromEnd(ListNode head, int n) {
            if(head.next==null && n == 1)return null;
            ListNode nxt = head;
            ListNode now = head;
            int k = n;
            while(k-->0){
                nxt = nxt.next;
            }
            if(nxt==null){
                head = head.next;
                return head;
            }
            while(nxt.next!=null){
                now = now.next;
                nxt = nxt.next;
            }
            now.next = now.next.next;
            return head;
        }
    }
}

File: go/P0019.go
//题意：删除倒数第n个节点
//题解：不用新建头节点之前的节点，因为如果删除头会在前面判断出来直接返回掉
package main

/*
 * Definition for singly-linked list.
 * type ListNode struct {
 *     Val int
 *     Next *ListNode
 * }
 */
 func removeNthFromEnd(head *ListNode, n int) *ListNode {
    slow, fast := head,head
    for i:=0;i<n;i++{
        fast = fast.Next
    }
    //当n是总数时，fast走完就是nil了，会导致下面报错
    if fast == nil{
        return head.Next
    }

    for fast.Next!=nil{
        slow,fast = slow.Next,fast.Next
    }
    slow.Next = slow.Next.Next
    return head
}